cli/src/commands/git/colocate.rs

```diff
@@ -0,0 +1,171 @@
+// Copyright 2024 The Jujutsu Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fs;
+
+use jj_lib::backend::Backend;
+use jj_lib::file_util::IoResultExt;
+use jj_lib::git;
+use jj_lib::repo::Repo;
+use tracing::instrument;
+
+use crate::cli_util::CommandHelper;
+use crate::command_error::internal_error;
+use crate::command_error::user_error;
+use crate::command_error::CommandError;
+use crate::commands::git::maybe_add_gitignore;
+use crate::ui::Ui;
+
+/// Make the current workspace colocated
+///
+/// This has a similar effect to `jj git init --colocate`, except
+/// it works on existing JJ repositories. The end result is a repo
+/// where you can run Git commands as well as JJ commands.
+///
+/// Limitations:
+///
+/// - Currently only works in the primary workspace.
+///
+/// - Your repo must have been initialized by `jj git init/clone`. Other
+///   configurations are not supported.
+#[derive(clap::Args, Clone, Debug)]
+pub struct GitColocateArgs {}
+
+#[instrument(skip_all)]
+pub fn cmd_git_colocate(
+    ui: &mut Ui,
+    command: &CommandHelper,
+    _args: &GitColocateArgs,
+) -> Result<(), CommandError> {
+    let workspace_command = command.workspace_helper(ui)?;
+    let workspace = workspace_command.workspace();
+
+    if workspace_command.working_copy_shared_with_git() {
+        return Err(user_error(format!(
+            "Workspace '{}' is already colocated.",
+            workspace.workspace_id().as_str()
+        )));
+    }
+
+    let Some(git_backend) = workspace_command.git_backend() else {
+        return Err(user_error(
+            "This repo is not using the Git backend, so you cannot colocate a workspace.",
+        ));
+    };
+
+    let dotgit_path = workspace.workspace_root().join(".git");
+    if dotgit_path.exists() {
+        return Err(user_error(format!(
+            "Path {} already exists, cannot colocate this workspace.",
+            dotgit_path.display()
+        )));
+    }
+
+    let git_repo = git_backend.git_repo();
+    if workspace.is_primary_workspace() {
+        let common_dir = git_repo.common_dir();
+        let old_path = workspace_command.repo_path().join("store").join("git");
+        let common_dir_canon = common_dir.canonicalize().context(common_dir)?;
+        let old_path_canon = old_path.canonicalize().ok();
+
+        let mut config =
+            gix::config::File::from_git_dir(common_dir.to_path_buf()).map_err(internal_error)?;
```

Comment:
nit: maybe this should be a user error (assuming it would do some I/O)?

BTW, the gix' fetch/clone implementation might help understand their config API. I'm not familiar with that.
https://github.com/GitoxideLabs/gitoxide/blob/main/gix/src/clone/fetch/util.rs